sky/serve/autoscalers.py

```diff
@@ -315,3 +342,159 @@ def _get_replica_ids_to_scale_down(num_limit: int) -> List[int]:
+            f'Number of ready spot instances: {num_ready_spot}, '
+            f'Number of alive on-demand instances: {num_launched_on_demand}, '
+            f'Number of ready on-demand instances: {num_ready_on_demand}')
+
+        scaling_options: List[AutoscalerDecision] = []
+        all_replica_ids_to_scale_down: List[int] = []
+
+        # Decide how many spot instances to launch.
+        num_spot_to_provision = (self.target_num_replicas -
+                                 self.base_ondemand_fallback_replicas)
+        if num_launched_spot < num_spot_to_provision:
+            # Not enough spot instances, scale up.
+            # Consult spot_placer for the zone to launch spot instance.
+            num_spot_to_scale_up = num_spot_to_provision - num_launched_spot
+            if self.spot_placer is not None:
+                locations = self.spot_placer.select(
+                    provisioning_and_launched_new_replicas,
+                    num_spot_to_scale_up)
+                assert len(locations) == num_spot_to_scale_up
+                for location in locations:
+                    spot_override = (
+                        self._get_spot_resources_override_dict(location))
+                    logger.info(
+                        f'Chosen location {location} with {self.spot_placer}')
+                    scaling_options.append(
+                        AutoscalerDecision(AutoscalerDecisionOperator.SCALE_UP,
+                                           target=spot_override))
+            else:
+                for _ in range(num_spot_to_scale_up):
+                    scaling_options.append(
+                        AutoscalerDecision(
+                            AutoscalerDecisionOperator.SCALE_UP,
+                            target=self._get_spot_resources_override_dict()))
+        elif num_launched_spot > num_spot_to_provision:
+            # Too many spot instances, scale down.
+            # Get the replica to scale down with select_replicas_to_scale_down
+            num_spot_to_scale_down = num_launched_spot - num_spot_to_provision
+            all_replica_ids_to_scale_down.extend(
+                RequestRateAutoscaler.select_replicas_to_scale_down(
+                    num_spot_to_scale_down,
+                    filter(lambda info: info.is_spot,
+                           provisioning_and_launched_new_replicas)))
+
+        # Once there is min_replicas number of
+        # ready new replicas, we will direct all traffic to them,
+        # we can scale down all old replicas.
+        all_replica_ids_to_scale_down.extend(
+            self.select_outdated_replicas_to_scale_down(replica_infos))
+
+        # Decide how many on-demand instances to launch.
+        num_on_demand_to_provision = self.base_ondemand_fallback_replicas
+        if self.dynamic_ondemand_fallback:
+            num_on_demand_to_provision += (num_spot_to_provision -
+                                           num_launched_spot)
```

Comment:
`num_spot_to_provision - num_ready_spot` this might not work; since spots take time to be ready, in that meantime, `num_on_demand_to_provision` will always be positive.